### PR TITLE
BDM elements in 3D

### PR DESCRIPTION
**Pull request by guidokanschat**
Implementation of BDM elements in three dimensions, including fixing the degrees of freedom on faces.

Make FETools::get_fe_from_name() work for BDM and fix code for ABF.

Final tests are still missing, but I thought I put it out for review already.

Thanks to @JaneTao for pointing out the error in the polynomials

### CODE REVIEW

File: source/fe/fe_bdm.cc [hunk 1]
```diff
@@ -266,94 +282,102 @@ FE_BDM<dim>::get_ria_vector (const unsigned int deg)
...
+  // This function sets up the values of the polynomials we want to
+  // take moments with in the quadrature points. In fact, we multiply
+  // thos by the weights, such that the sum of function values and
+  // test_values over quadrature points yields the interpolated degree
+  // of freedom.
+  template <int dim>
+  void
+  initialize_test_values (std::vector<std::vector<double> > &test_values,
+                          const Quadrature<dim> &quadrature,
+                          const unsigned int deg)
```
Comment (bangerth) [inline source/fe/fe_bdm.cc]: Would you be willing to put input arguments before output arguments?

File: source/fe/fe_bdm.cc [hunk 2]
```diff
@@ -266,94 +282,102 @@ FE_BDM<dim>::get_ria_vector (const unsigned int deg)
...
+  Quadrature<dim> faces = QProjector<dim>::project_to_all_faces(face_points);
+  for (unsigned int k=0; k < face_points.size()*GeometryInfo<dim>::faces_per_cell; ++k)
+    this->generalized_support_points[k]
+      = faces.point(k+QProjector<dim>
+                    ::DataSetDescriptor::face(0, true, false, false,
+                                              this->dofs_per_face));
+
+  // Currently, for backward compatibility, we do not use moments, but
+  // point values on faces in 2D. In 3D, this is impossible, since the
+  // moments are only taken with repsect to PolynomialsP.
```
Comment (bangerth) [inline source/fe/fe_bdm.cc]: repsect -> respect

**Comment by bangerth** [comment]
Nice! At the workshop today, someone mentioned that the BDM element "almost" works in 3d when removing some asserts and fixing a single line -- and showed that then the lowest order element works in 3d as well. Apparently, the higher order elements were more problematic, though. Maybe your patch addresses those as well.

Could you
* Add an entry to changes.h
* Add some kind of tests to verify the correctness now and in the future?

**Comment by tjhei** [comment]
I think @alistairbntl might be able to contribute some testcases.

File: source/fe/fe_bdm.cc (hunk 1 above)
Comment (guidokanschat) [inline source/fe/fe_bdm.cc]: No. I just got annoyed again about the output arguments in the middle of the argument list in FETools.

**Comment by guidokanschat** [comment]
Biggest issue were the node functionals on faces. Clearly code only suitable for 2D only.

@alistairbntl Test cases will be appreciated! I will just look into the standard finite element tests that have BDM and add 3D to them

**Comment by JaneTao** [comment]
talking about ABF spaces, since there's P_(k+2) involved, it became very delicate for k=0 case, where I have to increase QGauss order to degree+2 for the embedding matrix computation in FETools.

Before: QGauss(degree+1) leads to
2(degree+1) -1 >= degree+2
degree>=1

**Comment by guidokanschat** [comment]
@JaneTao That sounds like a bug in FE_ABF. FiniteElementData::degree is supposed to be the hightest order of a one dimensional polynomial needed to represent the element. Are you saying this is not it?

**Comment by guidokanschat** [comment]
Added 3D to all tests in fe with bdm in their name except fe_bdm_2.cc which is genuinely 2D.

Most of them just add output, but the ones inverting the mass matrix fail. Does anyone want to look into this issue?

By the way: using monomials for the additional shape functions might not be the best idea. Actually, using Legendre polynomials here and for PolynomialsP might be a much better idea and reduce condition numbers.

**Comment by guidokanschat** [comment]
Question: do we want to merge this state (after the changelog) so that more people can work on it?

**Comment by bangerth** [comment]
It seems like a step backward to merge a patch that breaks a test (even if a new one). Do you have an idea what happens? Did you check that at least some of the additional output is actually correct?

**Comment by guidokanschat** [comment]
I did check with my eigenvalue code and get convergence to the right eigenvalues. I do observe some numerical instabilities though on sone meshes which might be related to the mass matrix issue.

I worked on this for two days I did not have after I was asked to look into it. I have no capacity to look into the further details. I can help by answering questions, though.

My recommendation is looking at the spectrum of the cell mass natrix and see if it is indeed s. p. d.

If this pull request is indeed considered a "step back", I am also willing to withdraw it.

**Comment by alistairbntl** [comment]
I would be happy to volunteer some time to investigate this further, but I imagine I will require some advice as I proceed.  What is the best forum for asking questions / seeking advice when investigating a problem such as this?

**Comment by bangerth** [comment]
@alistairbntl -- talk to me sometime today.

@guidokanschat -- thanks for the eigenvalue checks, that seems sensitive enough. We'll take it from here.

I wanted to hit the "merge" button but there's a merge conflict. If you have time to resolve, can you do that? If not, drop me a line and I'll take care of it.

**Comment by guidokanschat** [comment]
I'll look into it later, which I am not sure will be before breakfast. Or, maybe as acting dean I can make someone do it.... ?

**Comment by guidokanschat** [comment]
Looks like I found it

**Comment by bangerth** [comment]
Nice, many thanks!

**Comment by tamiko** [comment]
We have a regression in 14 tests (*2 for debug) [1] mainly for BDM in 3D.

[1] http://cdash.kyomu.43-1.org/viewTest.php?onlydelta&buildid=10735

**Comment by guidokanschat** [comment]
Tests before my last rebase:

Test project /export/home/gkanscha/local/build/main
    Start 169: all-headers/base/memory_consumption.h.debug
1/4 Test #169: all-headers/base/memory_consumption.h.debug .....   Passed    0.10 sec
    Start 170: all-headers/base/memory_consumption.h.release
2/4 Test #170: all-headers/base/memory_consumption.h.release ...   Passed    0.08 sec
    Start 3512: deal.II/memory_consumption_01.debug
3/4 Test #3512: deal.II/memory_consumption_01.debug .............   Passed    0.16 sec
    Start 3513: deal.II/memory_consumption_01.release
4/4 Test #3513: deal.II/memory_consumption_01.release ...........   Passed    0.14 sec

This was definitely a surprise since it does not use anything BDM related. But the diffs to my final commit show a lot of work in FEValues etc. Are you sure you can blame this on the BDM change?

Wrong conjecture: here is the test of the latest state of the branch.

Test project /home/guido/build/clang
    Start  81: all-headers/base/memory_consumption.h.debug
1/4 Test  #81: all-headers/base/memory_consumption.h.debug .....   Passed    1.33 sec
    Start  82: all-headers/base/memory_consumption.h.release
2/4 Test  #82: all-headers/base/memory_consumption.h.release ...   Passed    1.33 sec
    Start 3671: deal.II/memory_consumption_01.debug
3/4 Test #3671: deal.II/memory_consumption_01.debug .............   Passed    1.88 sec
    Start 3672: deal.II/memory_consumption_01.release
4/4 Test #3672: deal.II/memory_consumption_01.release ...........   Passed    1.85 sec

**Comment by guidokanschat** [comment]
Now looking at the more probable ones

**Comment by bangerth** [comment]
deal.II/memory_consumption was broken by #1247. I'm working on this.

**Comment by tamiko** [comment]
@guidokanschat I was not terribly clear - not all failing tests on this commit have something to do with BDM :-D

**Comment by guidokanschat** [comment]
It seems that the commit with most of the updated test results got lost in a rebase. Good, as you will see soon

**Comment by bangerth** [comment]
Just for reference, the regression tester had these numbers of failures:
```
4a15113 	0	0	0     0 0 33
4a15113 	0	0	0     0 0 33
b377f2e 	0	0	0     0 0 33
dcb8693 	0	0	0     0 0 29
8185c2d 	0	0	0     0 0 27
8185c2d 	0	0	0     0 0 28
b87594d 	0	0	0     0 0 2
3d9c14a 	0	0	0     0 0 1
```
8185c2d was the merge of this PR.

**Comment by tamiko** [comment]
@bangerth see 8fa08ec for a first round of bugfixes and [1] for the result after that merge. There are still 4 tests failing.
[1] http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10780

**Comment by bangerth** [comment]
Yes, the ones that I still see failing on the 'tester' machine (http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10800) are these:
* fe/bdm_[1389]
* fe/fe_data_test

There are also find_cell_* which must somehow be unrelated (I imagine they are from the work on transform_real_to_unit_cell in #1177).

**Pull request merged**